src/main/java/com/fasterxml/jackson/databind/jsontype/impl/SimpleClassNameIdResolver.java

```diff
@@ -0,0 +1,194 @@
+            ConcurrentHashMap<String, String> typeToId,
+            HashMap<String, JavaType> idToType)
+    {
+        super(baseType, config.getTypeFactory());
+        _config = config;
+        _typeToId = typeToId;
+        _idToType = idToType;
+        _caseInsensitive = config.isEnabled(MapperFeature.ACCEPT_CASE_INSENSITIVE_VALUES);
+    }
+
+    public static SimpleClassNameIdResolver construct(MapperConfig<?> config, JavaType baseType,
+            Collection<NamedType> subtypes, boolean forSer, boolean forDeser)
+    {
+        // sanity check
+        if (forSer == forDeser) throw new IllegalArgumentException();
+
+        final ConcurrentHashMap<String, String> typeToId;
+        final HashMap<String, JavaType> idToType;
+
+        if (forSer) {
+            // Only need Class-to-id for serialization; but synchronized since may be
+            // lazily built (if adding type-id-mappings dynamically)
+            typeToId = new ConcurrentHashMap<>();
+            idToType = null;
+        } else {
+            idToType = new HashMap<>();
+            // 14-Apr-2016, tatu: Apparently needed for special case of `defaultImpl`;
+            //    see [databind#1198] for details: but essentially we only need room
+            //    for a single value.
+            typeToId = new ConcurrentHashMap<>(4);
+        }
+        final boolean caseInsensitive = config.isEnabled(MapperFeature.ACCEPT_CASE_INSENSITIVE_VALUES);
+
+        if (subtypes != null) {
+            for (NamedType t : subtypes) {
+                // no name? Need to figure out default; for now, let's just
+                // use non-qualified class name
+                Class<?> cls = t.getType();
+                String id = t.hasName() ? t.getName() : cls.getSimpleName(); // not {@code _defaultTypeId(cls);}
+                if (forSer) {
+                    typeToId.put(cls.getName(), id);
+                }
+                if (forDeser) {
+                    // [databind#1983]: for case-insensitive lookups must canonicalize:
+                    if (caseInsensitive) {
+                        id = id.toLowerCase();
+                    }
+                    // One more problem; sometimes we have same name for multiple types;
+                    // if so, use most specific
+                    JavaType prev = idToType.get(id); // lgtm [java/dereferenced-value-may-be-null]
+                    if (prev != null) { // Can only override if more specific
+                        if (cls.isAssignableFrom(prev.getRawClass())) { // nope, more generic (or same)
+                            continue;
+                        }
+                    }
+                    idToType.put(id, config.constructType(cls));
+                }
+            }
+        }
+        return new SimpleClassNameIdResolver(config, baseType, typeToId, idToType);
+    }
+
+    @Override
+    public JsonTypeInfo.Id getMechanism() { return JsonTypeInfo.Id.SIMPLE_CLASS_NAME; }
```

Comment:
SIMPLE_CLASS_NAME -> SIMPLE_NAME (also need PR for `jackson-annotations` for that)